### PR TITLE
Save all files in a task at the same time to avoid recomputing intermediate results

### LINKED ISSUE #2042: Add support for delayed saves
Since [iris 3.6](https://scitools-iris.readthedocs.io/en/stable/whatsnew/3.6.html), iris supports delayed saves. This is a great new feature that allows:
- checking that a preprocessor run will succeed before doing any data crunching, saving (compute) time if there is something wrong with the data
- faster preprocessor runs because
  - intermediate results can be reused to write multiple output files (e.g. regrid only once for multimodel statistics)
  - I suspect this will allow to better overlap computations with writing/reading data, but I did not measure it so far.

See #1714 for an example implementation.

### LINKED ISSUE #2521: `mask_fillvalues` and `mask_multimodel` recompute mask from all input cubes for every output cube
Because [`mask_fillvalues`](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/api/esmvalcore.preprocessor.html#esmvalcore.preprocessor.mask_fillvalues) and [`mask_multimodel`](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/api/esmvalcore.preprocessor.html#esmvalcore.preprocessor.mask_multimodel) are now lazy, they recompute the mask based on all the input cube for every output cube. This is slow and unnecessary because the mask is the same for every output cube.

### PR DESCRIPTION
**Pull request by bouweandela**
## Description

Save all files in a task at the same time to avoid recomputing intermediate results.

This change is not backward compatible because it changes the return value of [`esmvalcore.preprocessor.save`](https://esmvaltool--2522.org.readthedocs.build/projects/ESMValCore/en/2522/api/esmvalcore.preprocessor.html#esmvalcore.preprocessor.save), which is part of the public API. Previously this function returned the filename, not it returns `None` on immediate saves and a `dask.delayed.Delayed` for delayed saves that can be requested with the `compute=False` argument.

Closes #2521
Closes #2042 

Links to documentation:
- https://esmvaltool--2522.org.readthedocs.build/projects/ESMValCore/en/2522/quickstart/configure.html#logging-configuration
- https://esmvaltool--2522.org.readthedocs.build/projects/ESMValCore/en/2522/api/esmvalcore.preprocessor.html#esmvalcore.preprocessor.save

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [ ] [🛠][1] ~Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)~
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Review by valeriupredoi — CHANGES_REQUESTED** [review]
@bouweandela this is very nice, but I have serious concerns related to `store `and locking, there are many known issues with HDF5 and threads, and even the Dask folk are looking at this type of IO issue, see eg https://github.com/dask/distributed/issues/780 - inherently what iris are doing is also not thread-safe, so I am doubly concerned

**Comment by bouweandela** [comment]
> there are many known issues with HDF5 and threads, and even the Dask folk are looking at this type of IO issue, see eg https://github.com/dask/distributed/issues/780 -

The linked issue is not related to writing to HDF5 files.

> inherently what iris are doing is also not thread-safe, so I am doubly concerned

If you have concerns about Iris's capability, I would recommend playing around with it and see if you can get it to crash and report any issues you find on the Iris GitHub page. The code that handles saving to NetCDF with distributed lives in [`iris.fileformats.netcdf](https://github.com/SciTools/iris/tree/main/lib/iris/fileformats/netcdf) in case you would like to have a look.

> I have serious concerns related to store and locking

Could you elaborate on those and provide an example of a case where it does not work?

**Comment by valeriupredoi** [comment]
> Could you elaborate on those and provide an example of a case where it does not work?

I need to hatch me a few used cases and test for stress points. Not on the priority list though, so let's see if any issues pop up naturally, am not gonna block this PR, just wanted to see if you jave any concerns too :beer:

**Review by valeriupredoi — APPROVED** [review]

**Comment by valeriupredoi** [comment]
it seems that a Lock object is indeed in dask.distributed, so first hurdle I was afraid of is alleviated https://github.com/dask/dask/issues/1892#issuecomment-352942141

**Comment by schlunma** [comment]
Thanks Bouwe, looks good to me! I will test this with some real recipes and merge if successful 🚀

**Comment by schlunma** [comment]
I successfully tested this with many recipes with the default scheduler, a distributed `LocalCluster` and a distributed `SLURMCluster`. Everything looks and works as expected 🚀 

The only remaining comment I have is regarding the progress bar. This messes up the log files a lot, e.g.

```
2024-10-31 16:42:07,321 UTC [841493] WARNING /work/bd0854/b309141/micromamba/envs/esm/lib/python3.11/site-packages/iris/fileformats/netcdf/loader.py:422: IrisFactoryCoordNotFoundWarning: Unable to find coordinate for variable 'ps'
  warnings.warn(

^M[#################                       ] | 42% Completed |  7.4s2024-10-31 16:42:07,522 UTC [841207] INFO    Progress: 10 tasks running, 6 tasks waiting for ancestors, 68/84 done
^M[#################                       ] | 44% Completed |  7.5s^M[##################                      ] | 46% Completed |  7.6s^M[###################                     ] | 48% Completed |  7.7s^M[###################                     ] | 49% Completed |  7.8s^M[####################                    ] | 51% Completed |  8.0s^M[#####################                   ] | 53% Completed |  8.1s^M[#####################                   ] | 54% Completed |  8.2s^M[######################                  ] | 55% Completed |  8.3s^M[######################                  ] | 55% Completed |  8.4s^M[######################                  ] | 55% Completed |  8.5s^M[######################                  ] | 55% Completed |  8.6s2024-10-31 16:42:08,633 UTC [841493] WARNING /work/bd0854/b309141/micromamba/envs/esm/lib/python3.11/site-packages/iris/fileformats/netcdf/loader.py:422: IrisFactoryCoordNotFoundWarning: Unable to find coordinate for variable 'ps'
  warnings.warn(

^M[######################                  ] | 55% Completed |  8.7s^M[######################                  ] | 56% Completed |  8.8s^M[######################                  ] | 56% Completed |  8.9s^M[######################                  ] | 56% Completed |  9.0s^M[######################                  ] | 56% Completed |  9.1s^M[######################                  ] | 56% Completed |  9.2s^M[######################                  ] | 56% Completed |  9.3s^M[######################                  ] | 57% Completed |  9.4s^M[######################                  ] | 57% Completed |  9.5s^M[######################                  ] | 57% Completed |  9.6s^M[######################                  ] | 57% Completed |  9.7s2024-10-31 16:42:09,744 UTC [841479] INFO    Computing and saving data for task diag_assess_zhai_constraint_cmip5_x/cl
2024-10-31 16:42:09,751 UTC [841490] INFO    Computing and saving data for task diag_x_zhai_cmip5/cl
^M[######################                  ] | 57% Completed |  9.9s^M[                                        ] | 0% Completed |  1.0s^M[                                        ] | 0% Completed |  1.1s^M[#######################                 ] | 57% Completed | 10.8s^M[                                        ] | 0% Completed |  1.1s^M[                                        ] | 0% Completed |  1.2s^M[#######################                 ] | 58% Completed | 10.9s^M[                                        ] | 1% Completed |  1.2s^M[                                        ] | 1% Completed |  1.3s^M[#######################                 ] | 58% Completed | 11.0s^M[                                        ] | 1% Completed |  1.3s^M[                                        ] | 1% Completed |  1.4s^M[#######################                 ] | 59% Completed | 11.1s^M[                                        ] | 1% Completed |  1.4s^M[                                        ] | 1% Completed |  1.5s^M[#######################                 ] | 59% Completed | 11.2s^M[                                        ] | 1% Completed |  1.5s^M[                                        ] | 1% Completed |  1.6s^M[#######################                 ] | 59% Completed | 11.3s^M[                                        ] | 1% Completed |  1.6s^M[                                        ] | 1% Completed |  1.7s^M[#######################                 ] | 59% Completed | 11.4s^M[                                        ] | 1% Completed |  1.7s^M[                                        ] | 1% Completed |  1.8s^M[#######################                 ] | 59% Completed | 11.5s^M[                                        ] | 1% Completed |  1.8s^M[                                        ] | 1% Completed |  1.9s^M[#######################                 ] | 59% Completed | 11.6s^M[                                        ] | 1% Completed |  1.9s^M[                                        ] | 1% Completed |  2.0s^M[#######################                 ] | 59% Completed | 11.7s^M[                                        ] | 2% Completed |  2.0s^M[                                        ] | 2% Completed |  2.1s^M[#######################                 ] | 59% Completed | 11.8s^M[                                        ] | 2% Completed |  2.1s^M[                                        ] | 2% Completed |  2.2s^M[#######################                 ] | 59% Completed | 11.9s^M[                                        ] | 2% Completed |  2.2s^M[                                        ] | 2% Completed |  2.3s^M[#######################                 ] | 59% Completed | 12.0s^M[                                        ] | 2% Completed |  2.3s^M[                                        ] | 2% Completed |  2.4s^M[#######################                 ] | 59% Completed | 12.1s^M[                                        ] | 2% Completed |  2.4s^M[#                                       ] | 2% Completed |  2.5s^M[########################                ] | 60% Completed | 12.2s^M[#                                       ] | 2% Completed |  2.5s^M[#                                       ] | 2% Completed |  2.6s^M[########################                ] | 60% Completed | 12.3s^M[#                                       ] | 3% Completed |  2.6s^M[#                                       ] | 3% Completed |  2.7s^M[########################                ] | 60% Completed | 12.4s^M[#                                       ] | 3% Completed |  2.7s^M[########################                ] | 60% Completed | 12.5s^M[#                                       ] | 3% Completed |  2.8s^M[#                                       ] | 3% Completed |  2.8s^M[#                                       ] | 3% Completed |  2.9s^M[########################                ] | 60% Completed | 12.6s^M[#                                       ] | 3% Completed |  2.9s^M[#                                       ] | 3% Completed |  3.0s^M[########################                ] | 60% Completed | 12.7s^M[#                                       ] | 3% Completed |  3.1s^M[#                                       ] | 3% Completed |  3.1s^M[########################                ] | 60% Completed | 12.8s^M[#                                       ] | 3% Completed |  3.2s^M[#                                       ] | 3% Completed |  3.2s^M[########################                ] | 60% Completed | 12.9s^M[#                                       ] | 4% Completed |  3.3s^M[#                                       ] | 4% Completed |  3.3s^M[########################                ] | 60% Completed | 13.0s^M[#                                       ] | 4% Completed |  3.4s^M[##                                      ] | 5% Completed |  3.4s^M[########################                ] | 60% Completed | 13.1s^M[##                                      ] | 5% Completed |  3.5s^M[##                                      ] | 5% Completed |  3.5s^M[########################                ] | 60% Completed | 13.2s^M[##                                      ] | 5% Completed |  3.6s^M[##                                      ] | 5% Completed |  3.6s^M[########################                ] | 60% Completed | 13.3s^M[##                                      ] | 6% Completed |  3.7s^M[##                                      ] | 6% Completed |  3.7s^M[########################                ] | 60% Completed | 13.4s^M[##                                      ] | 6% Completed |  3.8s^M[##                                      ] | 6% Completed |  3.8s^M[########################                ] | 60% Completed | 13.5s^M[##                                      ] | 6% Completed |  3.9s^M[##                                      ] | 7% Completed |  3.9s^M[########################                ] | 61% Completed | 13.6s^M[##                                      ] | 7% Completed |  4.0s^M[##                                      ] | 7% Completed |  4.0s^M[########################                ] | 61% Completed | 13.7s^M[###                                     ] | 7% Completed |  4.1s^M[###                                     ] | 7% Completed |  4.1s^M[########################                ] | 61% Completed | 13.8s^M[###                                     ] | 7% Completed |  4.2s^M[###                                     ] | 7% Completed |  4.2s^M[########################                ] | 61% Completed | 13.9s^M[###                                     ] | 8% Completed |  4.3s^M[###                                     ] | 8% Completed |  4.3s^M[########################                ] | 61% Completed | 14.0s^M[###                                     ] | 8% Completed |  4.4s^M[###                                     ] | 8% Completed |  4.4s^M[########################                ] | 61% Completed | 14.1s^M[###                                     ] | 9% Completed |  4.5s^M[###                                     ] | 9% Completed |  4.5s^M[########################                ] | 61% Completed | 14.2s^M[####                                    ] | 10% Completed |  4.6s^M[####                                    ] | 10% Completed |  4.6s^M[########################                ] | 61% Completed | 14.3s^M[####                                    ] | 10% Completed |  4.7s^M[####                                    ] | 11% Completed |  4.7s^M[########################                ] | 61% Completed | 14.4s^M[####                                
```

In addition, it also sometimes produces overlapping lines in the terminal when the tool is run interactively (I guess caused by multiple parallel processes?):

```
2024-10-31 17:01:05,150 UTC [1258106] INFO    Computing and saving data for task map/tas
[                                        ] | 0% Completed |  0.9s2024-10-31 17:01:06,120 UTC [1258108] INFO    Computing and saving data for task timeseries/tas_global
2024-10-31 17:01:37,377 UTC [1258106] INFO    Successfully completed task map/tas (priority 1) in 0:00:32.497795
[####                                    ] | 11% Completed | 31.3s2024-10-31 17:01:37,493 UTC [1258080] INFO    Progress: 2 tasks running, 2 tasks waiting for ancestors, 1/5 done
```

I don't think there is an easy fix for that, so my suggestion would be to just remove the progress bar. Although I quite like that actually 😢

**Comment by valeriupredoi** [comment]
Can the progress bar not be piped to file?

**Comment by bouweandela** [comment]
Thanks a lot for testing and reviewing! As far as I can see, the progress bar does not end up in the log files, only in the stdout output that SLURM records. Would it help to disable the progress bar if `max_parallel_tasks` is larger than 1 to avoid overlapping progress bars?

**Comment by schlunma** [comment]
That's right! Sorry, I only checked the SLURM log and was just assuming that the `main_log.txt` will look the same.

I guess disabling it for `max_parallel_tasks > 1` would help for stdout, but not for the SLURM log. I am not sure if many people use the SLURM log (I do for example), but if many people do that, it would be nice to have a version that's actually readable. The current one is definitely not.

**Comment by bouweandela** [comment]
The SLURM log still looks OK-ish with `max_parallel_tasks: 1` if you use a viewer that renders the backspaces used to 'update' the progress bar in place, e.g. if you view the log with `cat` or `less -r`. Have you tried any of these options?

**Comment by bouweandela** [comment]
Indeed that does not work. So how do you suggest we proceed?

**Comment by schlunma** [comment]
I am not sure. The easiest solution would probably be to drop it entirely, and I cannot think of another solution at the moment...

**Comment by bouweandela** [comment]
Would it be an option to change `main_log.txt` so the log messages look the same as what gets written to stdout? Then you wouldn't need those slurm logs so much anymore.

**Comment by schlunma** [comment]
Yes, I think it would be a good idea to homogenize the stdout and log output. However, I don't think it's a very good idea to deliberately make slurm logs unusable.

Maybe we can ask the community for their opinion? @ESMValGroup/esmvaltool-developmentteam are you actively using slurm log files when you work with ESMValTool? Please react to this comment with a 👍 if "yes",👎 if "no". Thank you!

**Comment by jfrost-mo** [comment]
There are a few things you can do to check whether stdout is going to an interactive terminal or not, such as with the `tty` coreutils command.

We could only output the progress bar if we are connected to an interactive input.

**Comment by jfrost-mo** [comment]
For a non interactive use case like that, it may be desirable to do something that doesn't overwrite old output, as that doesn't always work in non-interactive systems.

For example, when attached to a non-interactive terminal you could print the percentage completion as
```
  1%
  2%
  3%
 ...
```
Or do what many data tranfer tools do and just print dots to show progress is happening.

**Comment by bouweandela** [comment]
Also a nice idea, but that would require implementing our own progress bars instead of the ones provided by Dask, which is more work than I would be willing to put into this.

**Comment by valeriupredoi** [comment]
> this, though, is a bit scary [dask/dask#2488](https://github.com/dask/dask/issues/2488)

let's see if this hits us, after we merge - no point holding up a performance imrpovement for something that's putative :+1:

File: esmvalcore/preprocessor/__init__.py
Comment (valeriupredoi) [reply esmvalcore/preprocessor/__init__.py:661]: BTW no need for notebooks, see https://stackoverflow.com/questions/64982455/format-dask-diagnostics-progress-bar

Comment (bouweandela) [reply esmvalcore/preprocessor/__init__.py:661]: `tqdm`s progress bars look nice, but didn't work for me when I tried to use them to track progress for the ESGF downloads a few years ago. The maintainers of that package are not very responsive to the issue either: https://github.com/tqdm/tqdm/issues/1272#issuecomment-1801439084. I would be happy to try again though.

Comment (valeriupredoi) [reply esmvalcore/preprocessor/__init__.py:661]: let's see if it does the trick, can't be any worse than that picture Manu posted in https://github.com/ESMValGroup/ESMValCore/pull/2522#issuecomment-2454546238 :rofl:

**Comment by bouweandela** [comment]
The conclusion from community consultation at the workshop is that we would like to display progress bars (also for downloads) and will provide a configuration option to disable them.

**Comment by bouweandela** [comment]
> For example, when attached to a non-interactive terminal you could print the percentage completion

@jfrost-mo I implemented this solution for the case where a progress bar is not desired.

**Comment by bouweandela** [comment]
@schlunma It is now possible to disable the progress bars by setting a value other than 0 for `log_progress_interval` in the configuration, e.g.

```
logging:
  log_progress_interval: 10s
```

Does it look better now?

File: doc/quickstart/configure.rst
Comment (schlunma) [inline doc/quickstart/configure.rst:285]: Maybe we should also mention how this can be used in the command line? E.g.

```bash
--logging='{"log_progress_interval": 2.0}'
```

This should go into the corresponding "Command line arguments" section further up.

File: esmvalcore/preprocessor/_dask_progress.py
Comment (schlunma) [inline esmvalcore/preprocessor/_dask_progress.py:209]:
```suggestion
        log_progress_interval = 10.0
```

I think the default should be more than 1s, otherwise this will fill the logs of complex recipes very quickly.

**Review by schlunma — COMMENTED** [review]
Thanks Bouwe, that look great! I really love the new progress bar ❤️ . It is also nicely rendered in the SLURM log now (I tried VSCode, `vim`, and `cat`):

grafik

The only "issue" I found is that with a distributed scheduler, the progress bar sometimes does not finish with "100%" (I didn't see this with the default scheduler):

grafik

One "issue" with the progress logger is that with the default scheduler, there is a new line character after the bar, this does not happen with a distributed scheduler:

grafik

I wonder if it would make sense to add an option to disable the progress logging completely? Maybe with `log_progress_interval=-1`?

Apart from that, I only have minor comments on the doc. Thank you!!

**Review by schlunma — APPROVED** [review]
Thanks Bouwe, tested this again and all looks as expected now! Cheers!

Will merge now since two people approved. :rocket:

**Pull request merged**
